### PR TITLE
8290324: Move atomic operations outside of os_xxx.hpp

### PR DESCRIPTION
**Pull request by iklam**
The os_xxx.hpp files inject extra methods/fields that are specific to atomic operations into the `os` class. However, the injected methods/fields are used only by a specific os/cpu combination. Therefore, they should not be inside the `os` class, which should contain only APIs that are used across platforms.

- For ports where the `atomic_copy64()` function is used in a single file, I moved it as an inline function in that file
- Otherwise it's moved to `atomic_<os>_<cpu>.hpp`
- The linux/arm port is a little more involved, but the new code should be a little cleaner than the old code.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290324](https://bugs.openjdk.org/browse/JDK-8290324): Move atomic operations outside of os_xxx.hpp

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [397ba800](https://git.openjdk.org/jdk/pull/9501/files/397ba800db37712a4178d409f903e3026d3efae5)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [0d8a150f](https://git.openjdk.org/jdk/pull/9501/files/0d8a150f8a3fdfcec920732bc8a39d7c0855fb78)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9501/head:pull/9501` \
`$ git checkout pull/9501`

Update a local copy of the PR: \
`$ git checkout pull/9501` \
`$ git pull https://git.openjdk.org/jdk pull/9501/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9501`

View PR using the GUI difftool: \
`$ git pr show -t 9501`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9501.diff">https://git.openjdk.org/jdk/pull/9501.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Seems okay - good cleanup.

I'm okay with it if Aarch64 folk are.

Thanks.

**Comment by kimbarrett** [comment]
It seems weird that for x86 we have `_Atomic_move_long` while for other platforms we have `atomic_copy64`.  It's also odd that we have 64bit implementation of the latter but not the former, in the sense that one would expect to have both or neither.

I think all of the 64bit implementations of `atomic_copy64` are identical.  I wonder if a better refactoring might be to have a shared atomic_copy64.[inline.?]hpp that has a shared 64bit implementation and uses includes platform files (using the usual macros for that) to get 32bit implementations when appropriate.

Maybe these could all be dealt with as followups.

Several of the `atomic_copy64` implementations are being moved to places where they are being given C linkage.  What's the rationale for that?

File: src/hotspot/os_cpu/bsd_zero/atomic_bsd_zero.hpp [hunk 1]
```diff
@@ -285,12 +285,31 @@ inline T Atomic::PlatformCmpxchg<8>::operator()(T volatile* dest,
...
+#endif
+}
+
 template<>
 template<typename T>
 inline T Atomic::PlatformLoad<8>::operator()(T const volatile* src) const {
   STATIC_ASSERT(8 == sizeof(T));
   volatile int64_t dest;
-  os::atomic_copy64(reinterpret_cast<const volatile int64_t*>(src), reinterpret_cast<volatile int64_t*>(&dest));
+  atomic_copy64(reinterpret_cast<const volatile int64_t*>(src), reinterpret_cast<volatile int64_t*>(&dest));
```
Comment (kimbarrett) [inline src/hotspot/os_cpu/bsd_zero/atomic_bsd_zero.hpp:312]: [pre-existing] Why do we have these casts when `atomic_copy64` takes `volatile void*`?  (There are more like this that I didn't bother to comment directly.)

File: src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp [hunk 2]
```diff
@@ -433,18 +433,23 @@ extern "C" {
...
+  inline void atomic_copy64(const volatile void *src, volatile void *dst) {
+    *(jlong *) dst = *(const jlong *) src;
+  }
+
   void _Copy_conjoint_jlongs_atomic(const jlong* from, jlong* to, size_t count) {
     if (from > to) {
       const jlong *end = from + count;
       while (from < end)
-        os::atomic_copy64(from++, to++);
+        atomic_copy64(from++, to++);
```
Comment (kimbarrett) [inline src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp:445]: If `atomic_copy64` was a shared API with platform-specific implementations we could eliminate all these copies of `_Copy_conjoint_jlongs_atomic`.

File: src/hotspot/os_cpu/linux_arm/atomic_linux_arm.hpp [hunk 3]
```diff
@@ -30,6 +30,40 @@
...
+  static int32_t  add_bootstrap(int32_t add_value, volatile int32_t *dest);
+
+  static int32_t  xchg_bootstrap(int32_t exchange_value, volatile int32_t *dest);
+
+  static int32_t  cmpxchg_bootstrap(int32_t compare_value,
+                                    int32_t exchange_value,
+                                    volatile int32_t *dest);
+};
+
+extern arm_atomic_funcs _arm_atomic;
```
Comment (kimbarrett) [inline src/hotspot/os_cpu/linux_arm/atomic_linux_arm.hpp:65]: Why is arm_atomic_funcs not an AllStatic class?  (And it has a non-conventional name.)

**Comment by iklam** [comment]
> It seems weird that for x86 we have `_Atomic_move_long` while for other platforms we have `atomic_copy64`. It's also odd that we have 64bit implementation of the latter but not the former, in the sense that one would expect to have both or neither.
> 
> I think all of the 64bit implementations of `atomic_copy64` are identical. I wonder if a better refactoring might be to have a shared atomic_copy64.[inline.?]hpp that has a shared 64bit implementation and uses includes platform files (using the usual macros for that) to get 32bit implementations when appropriate.
> 
> Maybe these could all be dealt with as followups.

For this PR, I just wanted to move code around but otherwise avoid making any actual changes. I'd prefer to do the further clean up in a follow up issue.

> Several of the `atomic_copy64` implementations are being moved to places where they are being given C linkage. What's the rationale for that?

That was unintentional. They just happened to be moved to immediate above the only function that uses them, and that function happens to have C linkage. Maybe I should explicitly mark these atomic_copy64 as `static inline` to be clear that whether they are C or C++ linkage is inconsequential?

File: src/hotspot/os_cpu/bsd_zero/atomic_bsd_zero.hpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/os_cpu/bsd_zero/atomic_bsd_zero.hpp:312]: To be honest I don't know :-)

File: src/hotspot/os_cpu/linux_arm/atomic_linux_arm.hpp [hunk 4]
```diff
@@ -25,27 +25,27 @@
...
-  int32_t  (*add_func)(int32_t add_value, volatile int32_t *dest);
-
-  int32_t  (*xchg_func)(int32_t exchange_value, volatile int32_t *dest);
-
-  int32_t  (*cmpxchg_func)(int32_t compare_value,
-                           int32_t exchange_value,
-                           volatile int32_t *dest);
+class ARMAtomicFuncs : AllStatic {
+public:
+  typedef int64_t (*cmpxchg_long_func_t)(int64_t, int64_t, volatile int64_t*);
```
Comment (dholmes-ora) [inline src/hotspot/os_cpu/linux_arm/atomic_linux_arm.hpp:36]: Is the need for typedef's due to making this a proper class? Unless these typedef's get used elsewhere they don't need to exist.

**Review by dholmes-ora — APPROVED** [review]
Still okay. But I would have resisted making non-move-related changes and leave that for future cleanups.

**Comment by iklam** [comment]
I have created a new JBS issue ( https://bugs.openjdk.org/browse/JDK-8290745 - Refactor atomic_copy64() functions ) to capture @kimbarrett's comments on the various problems with the atomic copying functions.

**Review by kimbarrett — APPROVED** [review]

**Comment by iklam** [comment]
Thanks to @dholmes-ora and @kimbarrett for the review. Passed build tiers 1 and 5.
/integrate

**Pull request closed without merging**